accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -263,6 +268,167 @@ The **flatten** function is modified as follows:
+
+At [this location](https://github.com/dart-lang/language/blob/65b8267be0ebb9b3f0849e2061e6132021a4827d/specification/dartLangSpec.tex#L15477)
+about synchronous non-generator functions, the text is changed as follows:
+
+```
+It is a compile-time error if $s$ is \code{\RETURN{} $e$;},
+$T$ is neither \VOID{} nor \DYNAMIC,
+and $S$ is \VOID, \code{\VOID?}, or \code{\VOID*}.
+```
+
+_Comparing to Dart before null-safety, this means that it is no longer allowed
+to "return void to null" in a regular function._
+
+At [this location](https://github.com/dart-lang/language/blob/65b8267be0ebb9b3f0849e2061e6132021a4827d/specification/dartLangSpec.tex#L15525)
+about an asynchronous non-generator function with future value type `$T_v$`,
+the text is changed as follows:
+
+```
+It is a compile-time error if $s$ is \code{\RETURN{} $e$;},
+$T_v$ is neither \VOID{} nor \DYNAMIC,
+and \flatten{S} is \VOID, \code{\VOID?}, or \code{\VOID*}.
+```
+
+_Comparing to Dart before null-safety, this means that it is no longer allowed
+to "return void to null" in an `async` function, nor to "return a void future
+to null"._
+
+The next sentence is changed as follows:
+
+```
+It is a compile-time error if $s$ is \code{\RETURN{} $e$;},
+\flatten{S} is not \VOID,
+$S$ is not assignable to $T$,
+and flatten{S} is not a subtype of $T$.
+```
+
+_Comparing to Dart before null-safety, this means that it is now allowed
+to return a future when the future value type is a suitable future; for
+instance, we can have `return Future<int>.value(42)` in an `async` function
+with declared return type `Future<Future<int>>`. Conversely, it is no longer
+allowed to return a `Future<dynamic>` or `FutureOr<dynamic>` when the future
+value type is `Future<U>` for some `U` which is not a top type, because it is
+no longer enough that `dynamic` is assignable to `U`._
+
+The dynamic semantics specified at
+[this location](https://github.com/dart-lang/language/blob/65b8267be0ebb9b3f0849e2061e6132021a4827d/specification/dartLangSpec.tex#L15597)
+is changed as follows, where `$f$` is the enclosing function with declared
+return type `$T$`, and `$e$` is the returned expression with static type `$S$`:
+
+```
+When $f$ is a synchronous non-generator, evaluation proceeds as follows:
+The expression $e$ is evaluated to an object $o$.
+A dynamic error occurs unless the dynamic type of $o$ is a subtype of
+the actual return type of $f$
+(\ref{actualTypes}).
+Then the return statement $s$ completes returning $o$
+(\ref{statementCompletion}).
+
+\commentary{%
+The case where the evaluation of $e$ throws is covered by the general rule
+which propagates the throwing completion from $e$ to $s$ to the function body.%
+}
+
+When $f$ is an asynchronous non-generator with future value type $T_v$
+(\ref{functions}),
+the return statement $s$ is desugared at compile time.
+In the definition, the first item whose requirement is satisfied is used,
+and $\Gamma$ denotes the typing environment of $s$:
+
+\begin{itemize}
```

Comment:
As an aside - I don't think we should be writing extensive latex in these markdown documents, it's unreadable.

Reply:
I wanted to make it obvious what the changes are in the case where the proposal is a concrete language specification update.

Do you prefer that we don't aim at concrete language specification updates at all (in which case there would be a need to write more context), or perhaps that we write updates for the language specification, but using pure markdown to indicate an approximation of the update?

Reply:
My goals are:
  - As much as possible, we have a single point of truth for the *changes* involved for a new language feature
  - Have the changes specified in a readable format

The first is why I really don't want to land some changes in the formal spec and some in the feature spec. 
  -  We could do away with feature specs, but I do find it useful having a single point of truth for all of the *new* things involved in a language feature.  
  - We could land changes in the formal spec, but then we'd have to point to long series of diffs.
  - We could maintain a CL with the accumulated changes to the spec, but then people have to read it in unformatted latex, and also it's tricky to see what the workflow is to land changes.

I don't claim to have a perfect solution here.  What I've been doing is using the feature spec, written in markdown, as a readable source of truth, with the intention that we then land, at some point all of the changes implied by the feature spec in the formal spec.

I'm open to other workflows.  One thing we could consider is combining them: maintain a CL with the latex changes, which we aim to be fully comprehensive, and have the feature spec be an abbreviated summary with pointers to the full language?

Reply:
I'll get some more feedback outside this PR, and rewrite as needed to markdown in a separate PR.

Reply:
I think the overall readability is lowered when the null-safety spec and similar documents contain updated snippets of text from the language specification along with permalinks to the existing text. On the other hand, the link plus the updated text is unambiguous.

I created language issue #947 to clarify this.